src/Builder.php

```diff
@@ -0,0 +1,735 @@
+    {
+        if (!$time) {
+            $time = new \DateTime('NOW');
+        }
+        return $this->with('exp', $time->format(\DateTime::ATOM));
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed footer.
+     *
+     * @param string $footer
+     * @return self
+     */
+    public function withFooter(string $footer = ''): self
+    {
+        $cloned = clone $this;
+        $cloned->cached = '';
+        $cloned->token = $cloned->token->withFooter($footer);
+        return $cloned;
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed footer,
+     * representing the JSON-encoded array provided.
+     *
+     * @param array $footer
+     * @return self
+     * @throws PasetoException
+     */
+    public function withFooterArray(array $footer = []): self
+    {
+        $encoded = \json_encode($footer);
+        if (!\is_string($encoded)) {
+            throw new EncodingException('Could not encode array into JSON');
+        }
+        return $this->withFooter($encoded);
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed 'iat' claim.
+     *
+     * @param \DateTime|null $time
+     * @return self
+     */
+    public function withIssuedAt(\DateTime $time = null): self
+    {
+        if (!$time) {
+            $time = new \DateTime('NOW');
+        }
+        return $this->with('iat', $time->format(\DateTime::ATOM));
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed 'iss' claim.
+     *
+     * @param string $iss
+     * @return self
+     */
+    public function withIssuer(string $iss): self
+    {
+        return $this->with('iss', $iss);
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed 'jti' claim.
+     *
+     * @param string $id
+     * @return self
+     */
+    public function withJti(string $id): self
+    {
+        return $this->with('jti', $id);
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed 'nbf' claim.
+     *
+     * @param \DateTime|null $time
+     * @return self
+     */
+    public function withNotBefore(\DateTime $time = null): self
+    {
+        if (!$time) {
+            $time = new \DateTime('NOW');
+        }
+        return $this->with('nbf', $time->format(\DateTime::ATOM));
+    }
+
+    /**
+     * Return a new JsonToken instance with a changed 'sub' claim.
+     *
+     * @param string $sub
+     * @return self
+     */
+    public function withSubject(string $sub): self
+    {
+        return $this->with('sub', $sub);
+    }
+
+
+    /**
+     * Return a new JsonToken instance, with the provided cryptographic key used
+     * to authenticate (and possibly encrypt) the serialized token.
+     *
+     * @param KeyInterface $key
+     * @param bool $checkPurpose
+     * @return self
+     * @throws PasetoException
+     */
+    public function withKey(KeyInterface $key, bool $checkPurpose = false): self
+    {
+        $cloned = clone $this;
+
+        if ($checkPurpose) {
```

Comment:
I refactored this a bit in the other PR, but since that'll need rewriting anyway might as well do it here:
I think there are a few instances of duplicated code in the `with*` style methods, where the corresponding `set*` method could be deferred to, e.g. you could write instead:

```php
public function withKey(KeyInterface $key, bool $checkPurpose = false): self
{
    return (clone $this)->setKey($key, $checkPurpose);
}
```

I think I caught the other instances in https://github.com/paragonie/paseto/pull/37/commits/18d7b01727e91bda9c070b9c0078a45bb80cf65d#diff-8f9c9046a027b435093568edd36eff23R678 if you wanted to enumerate them.